### ISSUE TITLE
Gps.locate uses stale ping responses from older pings if they are still in event queue.

### ISSUE THREAD

**Issue by Wojbie** [SquidDev-CC/CC-Tweaked#901]
### Minecraft Version

1.16.x

### Version

1.98.1

### Details

When running `gps.locate()` function multiple times in row one can accidentally cause it to give out incorrect coordinates due to it using older ping responses that were still in event queue. This happens due to fact that it will use first set of pings that give it valid coordinates and leftover pings form any extra gps hosts will stay in queue to be picked by another call.

Easiest way to visualize the bug is to run this code on pocket:
```lua
local a,b,c
while true do
 local x,y,z = gps.locate()
 if x ~= a or y ~= b or z ~= c then
  print(x,y,z)
  a,b,c = x,y,z
 end
end
```
Moving straight up generates this output.
![obraz](https://user-images.githubusercontent.com/5893536/130279032-a75e49c6-f9f8-4654-aca7-80b8cb9f8c5e.png)

Issue can be partially negated by adding code that clears out current events from queue example:
```lua
local a,b,c
while true do
 os.queueEvent("gpstest")
 os.pullEvent("gpstest")
 local x,y,z = gps.locate()
 if x ~= a or y ~= b or z ~= c then
  print(x,y,z)
  a,b,c = x,y,z
 end
end
```
But that don't resolve the issue as there is always possibility that there are still gps hosts that have not yet responded to old ping.

This code resolves issue in single run cases by waiting 0.05 s before sending first ping...
```lua
local a,b,c
while true do
 sleep(0.05)
 local x,y,z = gps.locate()
 if x ~= a or y ~= b or z ~= c then
  print(x,y,z)
  a,b,c = x,y,z
 end
end
``` 

But i am not sure if it is a proper solution as it feels hacky and I believe (but was unable to test) that multiple `gps.locate()` calls ran in parallel may still use stale data created by other one if timing happens to be bad...

I spend some time thinking if there is a backwards compatible way to update ping protocol to have unique ids for each ping to ensure that any locate will only listen to its own ping responses but sadly i have not found totally backwards compatible way..
One of solutions would be to move to a different gps protocol that sends some kind of PING table with unique id inside instead of "PING" string and have hosts still respond to both "PING" strings as well as new pings with tables of `{ x, y, z, ["id"]=pingID }`.

Here is another image showcasing the issue using plethora lines: 
![obraz](https://user-images.githubusercontent.com/5893536/130280546-961a1165-8f26-4a69-8484-21bf55bb1a21.png)
And after adding sleeps:
![obraz](https://user-images.githubusercontent.com/5893536/130280578-8ccbfbe6-04ae-4943-a757-865588d7372d.png)

**Comment by SquidDev** [comment]
> But that don't resolve the issue as there is always possibility that there are still gps hosts that have not yet responded to old ping.

Surely this bit isn't an issue as if they haven't _sent_ the message yet then the distance parameter will be correct? I don't know - haven't tested!

**Comment by Wojbie** [comment]
> 
> 
> > But that don't resolve the issue as there is always possibility that there are still gps hosts that have not yet responded to old ping.
> 
> Surely this bit isn't an issue as if they haven't _sent_ the message yet then the distance parameter will be correct? I don't know - haven't tested!

When i ran test for that version of code i had way less wrong offsets but they still happened. 
![obraz](https://user-images.githubusercontent.com/5893536/130289604-b42663ed-2438-4f4a-b0b8-77223921e4b6.png)

**Comment by Wojbie** [comment]
I spend some more time testing this issue and it seems that adding just simple `sleep(0.05)` at beginning of gps.locate() function code to make it run from clear-ish state does make it work well even in parallel case i was afraid of. So unless someone feels like we really need to make new gps protocol it would be indeed a solution?

**Comment by LoganDark** [comment]
> I spend some more time testing this issue and it seems that adding just simple `sleep(0.05)` at beginning of gps.locate() function code to make it run from clear-ish state does make it work well even in parallel case i was afraid of. So unless someone feels like we really need to make new gps protocol it would be indeed a solution?

This sounds like an queue-clearing strategy I like to use where I'll randomly generate an event name, `os.queueEvent` it, and then steamroll over all events until I pull my random event from the queue. The benefit of this is that it's completely synchronous as long as the queue is behaving as expected (i.e. no misbehaving emulator or HLOS).

The downside is that downstream programs do not generally expect API functions that they call to steamroll the event queue except in certain cases, but this _is_ one of those cases.

**Comment by fatboychummy** [comment]
I just had a rather random idea here. Rather than having something which flushes the queue, have the GPS broadcaster also broadcast the computer ID. This way, if the receiver receives multiple gps responses from the same computer, it can know to discard the first one (oldest) and keep the newest one.

It will not resolve all issues, but in theory it could at least help mitigate it.

**Comment by Wojbie** [comment]
I don't believe that would help. Look at examples. This is single computer making multiple requests quickly. Having unique id for each request would solve things but would not be backwards compatible and would meant we have to run 2 pararell systems.

**Comment by fatboychummy** [comment]
Would breaking backwards compat here be all that bad? I don't think I've heard of anyone shipping their own gps library, rather just using the builtin.

**Comment by Lupus590** [comment]
Could we have GPS hosts send two messages? The old message for compatibility and a new one that has expandability in mind.
